File: leetcode_notes/leetcode/0203.移除链表元素/0203-移除链表元素.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    // 203
    // Reference: cspiration
    public ListNode removeElements(ListNode head, int val) {
        // Time: O(n) Space: O(1)
        if(head == null) return head;
        head.next = removeElements(head.next, val);
        return head.val == val ? head.next : head;
    }
}

File: leetcode_notes/leetcode_cn/0203.移除链表元素/0203-移除链表元素.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        if (head == NULL) return NULL;

         ListNode *cur = head, *next = head->next;
         while (next != NULL) {
             if (next->val == val) {
                 cur->next = next->next;
                 delete next;
                 next = cur->next;
             } else {
                 cur = cur->next;
                 next = next->next;
             }
         }
         if (head->val == val) {
             cur = head;
             head = head->next;
             delete cur;
         }
         return head;
     }
};